catalog/app/containers/Bucket/PackageUpdateDialog.js

```diff
@@ -0,0 +1,933 @@
+    left: 0,
+    position: 'absolute',
+    right: 0,
+    top: 0,
+  },
+  progressContainer: {
+    display: 'flex',
+    position: 'relative',
+  },
+  progressPercent: {
+    ...t.typography.h5,
+    alignItems: 'center',
+    bottom: 0,
+    display: 'flex',
+    justifyContent: 'center',
+    left: 0,
+    position: 'absolute',
+    right: 0,
+    top: 0,
+  },
+  progressSize: {
+    ...t.typography.body2,
+    color: t.palette.text.secondary,
+    marginTop: t.spacing(1),
+  },
+}))
+
+function FilesInput({ input, meta, uploads, setUploads, errors = {} }) {
+  const classes = useFilesInputStyles()
+
+  const value = input.value || { added: {}, deleted: {}, existing: {} }
+  const disabled = meta.submitting || meta.submitSucceeded
+  const error = meta.submitFailed && meta.error
+
+  const onInputChange = input.onChange
+
+  const totalSize = React.useMemo(
+    () => Object.values(value.added).reduce((sum, f) => sum + f.size, 0),
+    [value.added],
+  )
+
+  const warn = totalSize > PD.MAX_SIZE
+
+  // eslint-disable-next-line no-nested-ternary
+  const label = error
+    ? errors[error] || error
+    : warn
+    ? `Total size of new files exceeds recommended maximum of ${readableBytes(
+        PD.MAX_SIZE,
+      )}`
+    : 'Drop files here or click to browse'
+
+  const pipeValue = useMemoEq([onInputChange, value], () => (...fns) =>
+    R.pipe(...fns, onInputChange)(value),
+  )
+
+  const onDrop = React.useCallback(
+    (files) => {
+      if (disabled) return
+      pipeValue(
+        R.reduce(
+          (acc, file) => {
+            const path = getNormalizedPath(file)
+            return R.evolve(
+              {
+                added: R.assoc(path, file),
+                deleted: R.dissoc(path),
+              },
+              acc,
+            )
+          },
+          R.__, // eslint-disable-line no-underscore-dangle
+          files,
+        ),
+      )
+    },
+    [disabled, pipeValue],
+  )
+
+  const rm = (path) => {
+    pipeValue(R.evolve({ added: R.dissoc(path) }))
+    setUploads(R.dissoc(path))
+  }
+
+  const del = (path) => {
+    pipeValue(R.evolve({ deleted: R.assoc(path, true) }))
+  }
+
+  const restore = (path) => {
+    pipeValue(R.evolve({ deleted: R.dissoc(path) }))
+  }
+
+  const handleAction = (handler, ...args) => (e) => {
+    e.stopPropagation()
+    if (disabled) return
+    handler(...args)
+  }
+
+  const revertFiles = React.useCallback(() => {
+    if (disabled) return
+    onInputChange(meta.initial)
+    setUploads({})
+  }, [disabled, onInputChange, setUploads, meta.initial])
+
+  const { getRootProps, getInputProps, isDragActive } = useDropzone({ onDrop })
+
+  const totalProgress = React.useMemo(() => getTotalProgress(uploads), [uploads])
+
+  const computedEntries = useMemoEq(value, ({ added, deleted, existing }) => {
+    const p1 = Object.entries(existing).map(([path, { size }]) => {
+      if (path in deleted) {
+        return { type: 'deleted', path, size }
```

Comment:
This `type` "deleted" and `deleted` key in `action` object are the same. It's not a coincidence, they are the same on a purpose. In this case it's better to use `const` to avoid mistypings (and for better IDE support, and for future Typescript):

```js
const DELETED = 'deleted'
// ---
return { type: DELETED }
// ---
const action = { [DELETED]: { handler: rm } }
```

Reply:
yep, and those names are also used by style rules and in destructuring patterns, where it's not very convenient to use constants. i don't think it's the real problem and i dont like this approach (using constants for strings e.g. `const NAME = 'name'`) in general, it seems more like a crutch. btw i believe typescript would help with this, bc it has constant string types (not sure how to call it properly)